nullability/v0.2/nullability-v0.2.graphql

```diff
@@ -0,0 +1,100 @@
+"""
+Indicates that a position is semantically non null: it is only null if there is a matching error in the `errors` array.
+In all other cases, the position is non-null.
+
+Tools doing code generation may use this information to generate the position as non-null:
+
+```graphql
+type User {
+    # email is semantically non-null and can be generated as non-null by error-handling clients.
+    email: String @semanticNonNull
+}
+```
+
+The `field` argument is the name of the field if `@semanticNonNull` is applied to an object type extension:
+
+```graphql
+# extend the base schema to make User.email semantically non-null.
+extend type User @semanticNonNull(field: "email")
+```
+
+`field` is used for client applications that do not control the base schema and must use type extensions.
+When used on a field definition, `field` must be null.
```

Comment:
Unrelated to the main change in this PR, but...

I'm guessing you considered this already, but it would be possible to let regular GraphQL schema validation enforce this invariant for us by making a separate directive for the case where you specify a field, and then have that directive _require_ a field be specified, and removing the argument from the field version.

As it stands we need ad-hoc validation that you _do_ specify a field when used on a type, and that you _don't_ specify a field when used on a field.

```graphql
directive @semanticNonNull(level: Int = 0) repeatable on FIELD_DEFINITION
directive @semanticNonNullField(field: String!, level: Int = 0) repeatable on OBJECT
```

That said, I can see that having two directives is also cumbersome.

Reply:
Makes sense. The less custom validation I have to do, the happier I am. `@semanticNonNullField(field: String!)` feels like a mouthful of "fields".  Alternative suggestions:

```graphql
extend type User @semanticNonNullField(name: "friends", level: 0)
extend type User @addSemanticNonNull(forField: "friends", level: 0)
```

Also thinking about this, we could also remove the `repeatable` from the regular `@semanticNonNull`:

```
directive @semanticNonNull(levels: [Int] = [0]) on FIELD_DEFINITION
# this ones needs to be repeatable for different fields. Which means we need validation that different directives for the same field do not conflict but 🤷 
directive @semanticNonNullField(name: String, levels: [Int] = [0]) repeatable on FIELD_DEFINITION
```

Reply:
Yeah, if we make levels plural, then we can remove the repeatable for the FIELD_DEFINITION variant. `@addSemanticNonNull` seem fine!

So the remaining validation we need to do:
1. The field/level(s) being referenced exist and are currently nullable
2. For `@addSemanticNonNull`, validate that the field exists
3. Validate type compatibility with interfaces

In theory you could end up with a type getting marked as semantically non null in multiple ways. Probably that's okay? As long as it does not confuse tools:

```graphql
type User @addSemanticNonNull(forField: "name") {
  name: String @semanticNonNull
}
```

Or 

```graphql
type User {
  friends: [User] @semanticNonNull(levels: [1,1,1,])
}
```

Reply:
Name nit: I find the word `add` to be a bit off for a directive because in a way all directives add something. `semanticNonNullField` works better for me but not a strong opinion.

Reply:
Agree directives are declarative and shouldn't start with a verb. I settled with `@semanticNonNullField(name: String, levels: [Int])` in https://github.com/apollographql/specs/commit/ba09b5017bda49502740ce36725e8d518f561e7a, let me know what you think.

Also nitpick but it should really be `@semanticNonNullPosition(field: String, levels: [Int])` but the spec uses field instead of position (or something else) in a lot of places so I kept `Field` for consistency even if it's a bit wrong.